[ARM] Have custom lowering for ucmp and scmp

AZero13 opened the pull request:
Limited to non-thumb1 for scmp at the moment, since there is no good way to do it.

AZero13 commented:
I wrote these by hand. Honestly the implementation came to me in a dream last night...

davemgreen commented on llvm/test/CodeGen/Thumb/ucmp.ll:
```diff
@@ -1,151 +1,290 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
-; RUN: llc -mtriple=thumbv7-apple-darwin %s -o - | FileCheck %s
+; RUN: llc -mtriple=thumbv6m-eabi %s -o - | FileCheck %s -check-prefix=THUMB1
+; RUN: llc -mtriple=thumbv7-apple-darwin %s -o - | FileCheck %s -check-prefix=THUMB2
```
You can remove mve and lob I think. It is hasV8_1MMainlineOps that determines whether csel is present.

AZero13 commented:
@davemgreen Done!

AZero13 commented:
Okay we need the mattr

davemgreen commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -802,6 +802,12 @@ ARMTargetLowering::ARMTargetLowering(const TargetMachine &TM_,
     setOperationAction(ISD::BSWAP, VT, Expand);
   }
 
+  if (!Subtarget->isThumb1Only() && !Subtarget->hasMVEIntegerOps())
```
hasV8_1MMainlineOps() would be what controls csel/csinc/etc, which I assume are the intructions that are trying to be guarded against?

AZero13 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -1628,6 +1634,10 @@ bool ARMTargetLowering::useSoftFloat() const {
   return Subtarget->useSoftFloat();
 }
 
+bool ARMTargetLowering::shouldExpandCmpUsingSelects(EVT VT) const {
```
Actually mve generates better code than the custom lowering, hence why it is enabled there, but when it is not, then yes the cost modeling this best matches what we are doing for that.

AZero13 committed:
[ARM] Have custom lowering for ucmp and scmp

Limited to non-thumb at the moment, but we can do this for i32 in 3 steps, using subs to set the flags initially.

davemgreen reviewed (approved):
Other than the comment below this LGTM.

AZero13 commented:
@davemgreen Done! Can you please merge?
